lib/ansible/modules/cloud/amazon/ec2.py

```diff
@@ -1617,17 +1619,16 @@ def main():
     if not HAS_BOTO:
         module.fail_json(msg='boto required for this module')
 
-    ec2 = ec2_connect(module)
-
-    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module)
+    try:
+        if module.params.get('region') or not module.params.get('ec2_url'):
+            ec2 = ec2_connect(module)
+        elif module.params.get('ec2_url'):
+            _, ec2_url, aws_connect_kwargs = get_aws_connection_info(module)
+            ec2 = connect_ec2_endpoint(ec2_url, **aws_connect_kwargs)
 
-    if region:
-        try:
-            vpc = connect_to_aws(boto.vpc, region, **aws_connect_kwargs)
-        except boto.exception.NoAuthHandlerFound as e:
-            module.fail_json(msg=str(e))
-    else:
-        vpc = None
+        vpc = connect_vpc(**aws_connect_kwargs)
```

Comment:
`aws_connect_kwargs` won't be defined at this point unless `ec2_url` is set, which is the unlikelier of the code paths